### DISCUSSION TITLE
Compiling raylib to statc, dll Implicit and dll Explicit (MinGW-w64)

### DISCUSSION THREAD

**Discussion by Axle-Ozz-i-sofT** [raysan5/raylib#2492] [General]
Hi Ray
I have had several days of looking at all raylib howtos and build options that I could find, including all the FAQs I could find here, so if I have missed a solution I hope you can forgive me :)
I am attempting to build the static and dll versions of the library. Building the static version seams OK, but I can't find the switch for building the dll version.
What I am attempting to achieve:
I have to be able to offer to other students a straight forward way of including raylib + additional modules such as raygui on both Windows and Ubuntu, as well as explain the set up and use of the library with C, FreeBASIC and Python3. I assume Py being a script interpreter will be using a dll version behind the bindings, so I need to be able to offer a simple and transportable copy/method of raylib as well as the required resources in both static and dll, and explain the use of static and dll versions.

I just want to focus on the Windows C requirements at the moment. I am working with Embarcadero's current version Dev-C++ with the TDM-GCC 9.2.0 release. I do not use Linux subsystems, MSYS2 or Cygwin etc and don't set a system path for GCC Bin as I have a few alternative GCC/MinGW installs. I can set a TEMP Path environment to GCC from a .cmd file if I need to run a make file.

My questions:
(I have the raylib-4.0.0_win32_mingw-w64.zip pre-compiled as well and will ask a question on that at the end)

- "mingw32-make PLATFORM=PLATFORM_DESKTOP" will build a static version with libraylib.a - How do I build the dll version?
- Does the libraylib.a also include raygui etc from the \extras directory by default?
- Does the self compiled libraylib.a "mingw32-make PLATFORM=PLATFORM_DESKTOP" contain the raygui styles resources found in the separate project under \raygui-3.1\styles\?
- The same questions with regard to the inclusion of the extra resources applies to the pre-compiled versions libraylib.a, libraylibdll.a and raylib.dll as I notice that it has the extra includes raygui.h, raymath.h, rlgl.h. Are the alternative styles included in the pre-compiled .a and .dll or are they included in a binary array in the header, or do I have to still have to make the \raygui-3.1\styles\ available?

Ultimately I am just attempting to work out what minimum package is required to be place into the project source directory or into the Mingw includes/lib directories. and weather to just use the "raylib-4.0.0_win32_mingw-w64.zip" without any other requirements.

Sorry if I sound confusing with the above. if you need me to clarify my questions just ask and I will attempt to do better :)
Axle

**Comment by Axle-Ozz-i-sofT** [comment]
Update:
I have just discovered that the raylib raygui.h is different to raygui raygui.h which is different to "https://github.com/raysan5/raygui/blob/master/src/raygui.h'.
I have 
RAYGUI_ICON_NONE
RICON_NONE 
ICON_NONE

I was attempting to compile "raygui-3.1\examples\controls_test_suite\controls_test_suite.c" using "raylib-4.0.0_win32_mingw-w64.zip" linker: lib/libraylib.a
-lopengl32
-lgdi32
165	72	.\controls_test_suite.c	[Error] 'RAYGUI_ICON_FILE_SAVE' undeclared (first use in this function); did you mean 'RICON_FILE_SAVE'?

So we have multiple incompatible raygui.h ? And I will have to write a wrapper, or change the source depending if I am using raylib raygui.h, or raygui raygui.h ?
or
Do I just abandon the extras such as raygui source, and only use the raygui.h in raylib and change the examples from raygui to reflect the reflect the header in raylib?

Update 2:
I tried compiling the example controls_test_suite.c with raylib pre-built raylib.a
Corrected RAYGUI_ICON_FILE_SAVE to RICON_FILE_SAVE etc
Received the following error
\x86_64-w64-mingw32\bin\ld.exe	lib/libraylib.a(rcore.o):rcore.c:(.text+0x17d3f): undefined reference to `timeEndPeriod'

I tried including Windows.h (for timeapi.h) but then got more conflicts...
[Error] 'Rectangle' redeclared as different kind of symbol
where multiple windows.h are redeclared.

Is it just me not setting the includes and linker properly, or is raylib (pre-built bins) internal raygui.h incomparable with anything from the raygui-3.1 including the examples (ツ)

**Comment by raysan5** [comment]
Hi @Axle-Ozz-i-sofT,

 > "mingw32-make PLATFORM=PLATFORM_DESKTOP" will build a static version with libraylib.a - How do I build the dll version?

`mingw32-make PLATFORM=PLATFORM_DESKTOP RAYLIB_LIBTYPE=SHARED` should compile the `raylib.dll`.

 > Does the libraylib.a also include raygui etc from the \extras directory by default?

No. Adding `RAYLIB_MODULE_RAYGUI=TRUE` to make line should add it... but I haven't tested it in some time...

 > Does the self compiled libraylib.a "mingw32-make PLATFORM=PLATFORM_DESKTOP" contain the raygui styles resources found in the separate project under \raygui-3.1\styles?

No, styles are loaded by the user at runtime, should be added to user project.

 > The same questions with regard to the inclusion of the extra resources applies to the pre-compiled versions libraylib.a, libraylibdll.a and raylib.dll as I notice that it has the extra includes raygui.h, raymath.h, rlgl.h. Are the alternative styles included in the pre-compiled .a and .dll or are they included in a binary array in the header, or do I have to still have to make the \raygui-3.1\styles\ available?

Same answer as before, styles are added to the user project as separate modules.

 > Do I just abandon the extras such as raygui source, and only use the raygui.h in raylib and change the examples from raygui to reflect the reflect the header in raylib?

The `raygui` version in `raylib/src/extras` is outdated, the lat version is the one provided in https://github.com/raysan5/raygui. Actually, I'm considering removing all the `extras` directory to avoid this kind of confusions.

**Reply by Axle-Ozz-i-sofT** [reply]
Hi ray
thank you for the reply :)

mingw32-make PLATFORM=PLATFORM_DESKTOP RAYLIB_LIBTYPE=SHARED
I found some references to RAYLIB_LIBTYPE=SHARED in a Linux make file in \examples\Makefile
Do you have a document for the different compile options?

And looking at the source you are working from a Linux platform. Is this a Linux subsystem for Windows such as MSYS2? If so which one are you using? (Just so I can compare with native Windows builds)
(Everyone seams to be using Linux command line toolsets on windows now so it is getting difficult to find online help for working with native windows when compiling libraries)

I will work with the raylib/src and ignore the raylib/src/extras/raygui.h and use the separate raygui\raygui-3.1\src for GUI :)
The other 3 headers raylib/src/extras/ easings.h, physac.h and rmem.h are OK to continue using I assume?

Looking forward to getting raylib up and running and jumping into the API :P
I have an extremely busy few weeks ahead of me, so I am trying to get a clean set up and some project spaces so I can get familiar during some short break periods :)

regards
Axle

P.S. I am 55 and have been coding since I was 12, but I am not an engineer, so everyday is still a learning experience :)

**Reply by orcmid** [reply]
> @Axle-Ozz-i-sofT And looking at the source you are working from a Linux platform. Is this a Linux subsystem for Windows such as MSYS2? If so which one are you using? (Just so I can compare with native Windows builds) (Everyone seams to be using Linux command line toolsets on windows now so it is getting difficult to find online help for working with native windows when compiling libraries)

I think there is some confusion here.  I don't think you need MSYS2 although for all I know mingw32 uses it.

The Linux Subsystem for Windows is a kind of virtual machine that is near-native in Windows 10 and later.  So it runs native Linux distributions and has features that provide for more integration between the WSL and the Windows PC desktop, file system, and apps.

My approach to making native applications for Windows is to use the free Visual Studio Build Tools combined with VS Code as an ideal editor.   There are examples of that in the `raylib/projects/scripts/build-windows.bat`.  It works better to do static binding (which means the compiler settings are compatible across all the modules).  That won't help for Python and other programming systems, but it may take more than a direct DLL to get where you want.  I have been promising to provide a better, simpler version of `build-windows.bat` and I should simply get it done.

 That might not align with your ambitions for your students.  If you want them to be using Linux, even if on a PC, I think you need to use WSL and install and use raylib inside WSL, not using the Windows install of raylib.

> @Axle-Ozz-i-sofT  P.S. I am 55 and have been coding since I was 12, but I am not an engineer, so everyday is still a learning experience :)

PS: I'm a bit like you, though I didn't write my first lines of code until I was 19.  But that was in 1958 :).  I do consider myself a computer scientist, having grown up in the field.

**Reply by Axle-Ozz-i-sofT** [reply]
Hi orcmid
nah, I am fully aware of WSL and it is an additional windows install module, so not included by default. I DON'T use any of the Linux subsystems WSL, MYS2 or cygwin etc which is where it becomes a bit problematic as the library appears to be built using one of the Linux subsystems for Windows. Most of the examples and makefiles are using Linux paths ./mydir/ instead of .\mydir\ <- W10 onward does have some native support for Linux paths in the kernal but is not guaranteed without installing a Linux subsystem.

I use the native MinGW-w64 compiler set and library usually with Dev-C++ as the IDE. Code::Blocks was modeled of of it and has many similarities. I don't set a system %PATH% for the GCC compiler as I usually have multiple versions floating around, and ether set it from a cmd using set or via the IDE. Setting up the GCC compiler switches, Lib paths and linker -lmylib in the IDE is much the same between Dev-c++ and Code::Blocks and they used to share the same Library package manager. The lib packages haven't been updated for some time so it's a manual install task for additional "updated" libraries but usually it is straight forward to do :)

I keep all Linux programming separate and do that directly in Ubuntu or Lubuntu. <-Linux is a bit better/easier than Windows for dealing with library management.

I can code in most any C or BASIC like syntax language they want to throw at me. I started out on the 6502 Ohio Scientific C1P and C4P (aka Superboard-II) and progressed from there. My first words I saw was "Copyright Microsoft ..." so I have been building Ms windows systems most of my life. Although I have used Linux in the past I have only just recently (12 months) jumped into the Linux system and programming side. I built my own Linux distro that turned out (By accident) to be very similar to Lubuntu. So that was a nice learning experience :)

My lecturer is a 20+ years engineer and great mentor and throws heaps of worthy challenges at me. We just wrote up the programming modules(units) and guide for programming in Python together for QLD AU CertIII, so it was an enjoyable experience to work along side of him :)
This current task I am working is an extension of that as an additional information guide to programming.

Cheers
Axle

**Reply by Axle-Ozz-i-sofT** [reply]
@orcmid @raysan5 
I rechecked myself on the paths syntax and GCC will allow Linux paths as the portable option "C:/mydir/nextdir/file.ext" as opposed to Windows C:\\mydir\\nextdir\\file.ext"
Because I jump around so many different languages and have been caught out with the Unix paths so many time I set "Myself" a rule to use Windows Paths across the board.

So I apologize for my previous assertion.
Axle

**Reply by Axle-Ozz-i-sofT** [reply]
@raysan5
Hi ray
Still attempting to compile the library from source...
When I compile raylib.a using raylib\src\Makefile I get a smaller raylib.a than the pre-compiled raylib-4.0.0_win64_mingw-w64 (Using w64devkit from the install.exe) so I guess I am missing something?
I also made some changes in "raylib\src\Makefile" to attempt to include raygui.h so I could attempt a single header library file:
RAYLIB_MODULE_RAYGUI ?= TRUE
RAYLIB_MODULE_PHYSAC ?= TRUE
But the build fails due to a lack of raygui.c

`C:\raylib\raylib\src>mingw32-make
echo #define RAYGUI_IMPLEMENTATION > raygui.c

echo #include "../src/extras/raygui.h" >> raygui.c

gcc -c raygui.c -Wall -D_DEFAULT_SOURCE -Wno-missing-braces -Werror=pointer-arith -fno-strict-aliasing -std=c99 -s -O1 -
Werror=implicit-function-declaration -I. -Iexternal/glfw/include -Iexternal/glfw/deps/mingw -DPLATFORM_DESKTOP -DRAYGUI_
IMPLEMENTATION
cc1.exe: fatal error: raygui.c: No such file or directory
compilation terminated.
make: *** [Makefile:585: raygui.o] Error 1`

**Q1. Am I using the correct makefile to build the static or shared libraries?**
Q2. Is it possible to add the extras such as raygui for a single lib.a or dll.a dll include? Or am I going beyond the scope of the library's intended use?

The main goal here is to also build the bindings for FreeBASIC as the most recent raylib.bi doesn't include raygui.bi
https://github.com/WIITD/raylib-freebasic
Other versions of the bindigs are raylib  3.5

The PY module raylib-dynamic 4.0.0.6 CFFI appears to be up to date.
https://github.com/electronstudio/raylib-python-cffi

**Comment by Axle-Ozz-i-sofT** [comment]
Update 3:
I have attempted to use both raylib and raygui in the project source directory using the pre-compile MinGW-w6 lib a static as well as compiling directly from the source raylib-4.0.0 and  raygui-3.1 in an attempt to compile "raygui-3.1\examples\controls_test_suite\controls_test_suite.c" and continue to get the error "libraylib.a(rcore.o):rcore.c:(.text+0x1ceb0): undefined reference to `timeBeginPeriod'".
I have moved the example controls_test_suite.c to the root of my project folder with:
Raylib_tests
   \raygui-3.1\*
   \raylib-4.0.0\*

I changed the #includes in controls_test_suite.c to point to the paths of raylib.h and raygui.h

I have had a look at :
`        #if defined(SUPPORT_WINMM_HIGHRES_TIMER) && !defined(SUPPORT_BUSY_WAIT_LOOP)
            // NOTE: Those functions require linking with winmm library
            unsigned int __stdcall timeBeginPeriod(unsigned int uPeriod);
            unsigned int __stdcall timeEndPeriod(unsigned int uPeriod);
        #endif`

I am not sure why it is not defining timeBeginPeriod() ? :/
I had originally tryed including windows.h to get the timapi.h where it is defined, but then realized it is defined in rcore.c after all of the redefined errors lol

I did original compile this example using NP++ with your "w64devkit" from the install version and it worked fine, so not sure what I am missing?

**Reply by Axle-Ozz-i-sofT** [reply]
Update.. I was scraping the net trying to find the correct linker objects and missed -lwinmm :/
So now I am running with Linker:
`ray/src/libraylib.a
-lopengl32
-lgdi32
-lwinmm`
 An seams to be compiling OK.

And controls_test_suite.c runs as it should... Success! :)

^ Ahh, found it in the install version under NP++ npp\plugins\Config\npes_saved.txt

I think when I get my head around this I will create a separate indexed help document for working in native windows with Dev-C++ and Code::Blocks over the standard Mingw-w64 or the TDM-Mingw-w64 which I am using at the moment. <- If I do good with it I will send you a copy :)

Nice to have it working. I'll clean up and spend some time in the background with the examples and write up a few of my own to get familiar with the API.

Thanks for the help so far. It is much appreciated.

**Comment by Axle-Ozz-i-sofT** [comment]
Hi Ray
Not sure if I should start a new thread for each of these questions as I'm going off topic from the original Dll build title?

In the raylib.h ~line 2306 I am getting a compiler warning for the use of a binary literal. As this is C++14 and not C99 std can I change this to hex to keep to the standard and suppress the warning...
`while (((unsigned char)text[textLength - 1 - charUTF8Length] & 0b01000000) == 0) charUTF8Length++;
to
while (((unsigned char)text[textLength - 1 - charUTF8Length] & 0x40) == 0) charUTF8Length++;`
or will that mess with deleting the UTF-8 characters? (I'm sure the bitwise operator & should work the same with hex 0x being the native binary format for C99)

**Comment by Axle-Ozz-i-sofT** [comment]
Update:
I have managed to combine the pre-compiled "raylib-4.0.0_win64_mingw-w64" with "raygui-3.1" headers and styles (Replacing the existing raygui.h in .\include\

This appears to be working well with the examples and changing the styles and fonts, so I think I am on the right pathway :)
So I think I can include and transport it in a simple way, most likely in the project directory, but I will look at installing it directly to the Mingw root in \include and \lib.
You can view the project folder structure below :)

https://sites.google.com/view/axlestudent-ict/tafe/raylib-raygui-tests

**Comment by raysan5** [comment]
@Axle-Ozz-i-sofT So, is it working as expected right now?

Please note that you can access raylib Discord to ask about raylib building and configuration, there is a special channel for that and a big community always ready to help.

**Reply by Axle-Ozz-i-sofT** [reply]
@raysan5 "raylib Discord" I will have to have a look and see if I still have a Discord account. I tend to live in IRC most of the time :)

Yes, I have the Static builds set up in a convenient way as per the Snap2HTML output of the project directory. As I still have to do a static compile of the other included headers I will likely just stay with the static raylib.a and use "raylib-4.0.0_win64_mingw-w64" with "raygui-3.1" replacing the outdated \extras\raygui.h
The examples from both libraries are working fine :)

Likely I will (with your permission) when the time comes offer the students a copy of the compact "raylib-4.0.0_win64_mingw-w64_Ex" as a drop in for their projects folder as well as the links to the original source and binaries etc

I am currently attempting to build the FreeBASIC binders for the 2 most recent version 4.0 and 3.1. (The most recent available for FB is raylib 3.5 and raygui 2.8).
https://github.com/WIITD has created a partial binder for raylib 4.0, but no raygui at this time.

Python binders from PyPi are up to date with the latest versions.

Not sure how much difference their is between the raylib v3.5 and v4.0 but you show some significant changes in the change logs, so I would like to keep the same version across all 3 languages if I can :)

Thanks again for the response. I prob wont annoy you with many (if any) questions from here on. I'll let you know when I have completed these books with the raylib/gui examples :)

Again thank you for the assist :)
Axle

P.S. I have altered the binary constant from above 0b01000000 to HEX 0x40 as a permanent change.. I will test it with UTF-8 when I do the linux tests :)

**Comment by miguelnkono** [comment]
please I need help to compile raygui with mingw

**Reply by Axle-Ozz-i-sofT** [reply]
Hi [miguelnkono](https://github.com/miguelnkono)
The raylib team may have a better response for your question.
I am a little out of touch with the current raylib raygui builds but can attempt to assist.

Can you offer a little more information as to your difficulty compiling raygui?

Regards
Axle

**Comment by miguelnkono** [comment]
Thank you for the help but it's ok, I've already find a solution on how I
can compile it! But now, please I need a reference on how to begin with
since there's no doc for it. Please, looking forward for a pleasant reply
from you

**Reply by Axle-Ozz-i-sofT** [reply]
Most raylib raygui documents are found within the actual source code repository and the wiki.
https://github.com/raysan5/raylib/tree/master?tab=readme-ov-file#build-and-installation
Also look at the Makefile or CMake files depending upon the build system you are using.

Depending upon you OS, Compiler build chain (MSVC/GCC) etc. there may be some different methods required. This will also depend upon if you require static libraries (lib.lib/lib.a or dynamic libraries (lib.dll/lib.so).

Remember that raylib is the main rendering library and raygui is an add-on for raylib. The raygui guide should explain the different ways you can build raylib with raygui included or stand alone.

Good luck :)
Axle

**Reply by Axle-Ozz-i-sofT** [reply]
P.S. I am uncertain about how to compile raylib/raygui directly within a Code::Blocks Dev-C++ or MSVC studio project.
We usually compile the raylib/gui source into byte code an then add the compiled archive or dll/so and header file to the IDE project :)

**Reply by Axle-Ozz-i-sofT** [reply]
P.S.S.
For using the library API read the header file rayli.h and raygui.h
Also see the examples: (The controls_test_suite.c will give you a good overview)
https://github.com/raysan5/raylib/tree/master/examples
https://github.com/raysan5/raygui/tree/master/examples
https://www.raylib.com/examples.html
And the cheat sheet for the API
https://www.raylib.com/cheatsheet/cheatsheet.html

If that helps :)

**Reply by Axle-Ozz-i-sofT** [reply]
P.S.S.
I do have a book and guide on compiling raylib/raygui as well as setting it up in Dev-C++ and Code::Blocks on Windows and Ubuntu and compiling the examples but it is still in draft and not yet released.
The guide is also based upon an older version of GCC using sjlj instead of the more current x64 seh error handling.
Being draft the book still needs a little more included as well as some examples of the more recent GCC & raylib builds.
>
If you have specific question or issues on getting the project working in your IDE don't be afraid to ask as it can be a little confusing working out the build options, compiler flags, and the linker includes.
It's very OS and IDE specific, so be sure to include OS, IDE, Compiler version (ie.GCCv9.xx) and the MinGW build chain (ie. TDM GCC9.2.0 sjlj or gcc version 13.2.0 (MinGW-W64 x86_64-msvcrt-posix-seh, built by Brecht Sanders, r5)) when asking question so we can narrow down suggestions :)
Best regards
Axle

**Comment by miguelnkono** [comment]
Awesome, thanks for the info!  It'll check it out

**Comment by miguelnkono** [comment]
Hello, please I want to know if Raylib provide a way for accessing the file
system of the computer because I am building an audio player with it but
I'm stuck on this point.
Please could you help me with this?
Thank you.
